### PR TITLE
Raise IRQL to DISPATCH _after_ test_helper is successfully initialized.

### LINKED ISSUE #2973: Workflow failed - fault_injection_full
[Failed Run](https://github.com/microsoft/ebpf-for-windows/actions/runs/6492969559)
[Codebase](https://github.com/microsoft/ebpf-for-windows/tree/6f4bcbd00d2d803aa24e25f78d63f980175f524f)
Test name - `fault_injection_full`

### PR DESCRIPTION
**Pull request by dv-msft**
## Description

Fixes in this PR:
1. Raise IRQL _after_ `test_helper` initialization, so that its destructor is called at PASSIVE in we hit an error in `test_helper::initialize()`.
_(The `test_helper::~test_helper()` destructor calls `ebpf_flush_epoch()` that ends up blocking on a lock if called at DISPATCH.)_

2. Ensure that the `_ebpf_epoch_cpu_table` ptr is reset to NULL if `ebpf_epoch_initiate` frees that pointer in its error handling path.
_(Not doing so led to the 'use-after-free' crash seen in [this](https://github.com/microsoft/ebpf-for-windows/issues/2973#issuecomment-1759363487) comment, where the code path initiated by `ebpf_epoch_flush` ends up referencing this (now freed) pointer.)_

## Testing

_Do any existing tests cover this change? Are new tests needed?_
Issued exposed by existing test.  No new test needed.

Fix Verification:
Successful local run of:
`powershell.exe .\Test-FaultInjection.ps1 D:\wrk\ebpf-for-windows\x64\Debug 3600 .\unit_tests.exe 16`

## Documentation

_Is there any documentation impact for this change?_
No doc impact.

## Installation

_Is there any installer impact for this change?_
No installer impact.

Fixes #2973

### CODE REVIEW

**Comment by rectified95** [comment]
Since multiple failures can be placed under the same failing job while the issue is open, just making sure the PR addresses this specific callstack: https://github.com/microsoft/ebpf-for-windows/issues/2973#issuecomment-1759363487

If it's the same issue, could you quickly explain why this
> (The test_helper::~test_helper() destructor calls ebpf_flush_epoch() that ends up blocking on a lock if called at DISPATCH.)

causes a crash with the stack from my comment? :)

File: libs/runtime/unit/platform_unit_test.cpp [hunk 1]
```diff
@@ -1208,11 +1208,11 @@ has_dominant_frequency(size_t sequence_length, std::function<uint32_t()> random_
 
 TEST_CASE("verify random", "[platform]")
 {
-    KIRQL old_irql = KeRaiseIrqlToDpcLevel();
-
     _test_helper test_helper;
     test_helper.initialize();
 
+    KIRQL old_irql = KeRaiseIrqlToDpcLevel();
```
Comment (Alan-Jowett) [inline libs/runtime/unit/platform_unit_test.cpp:1214]: Can we make this lower on scope exit?

**Comment by dv-msft** [comment]
> Could you quickly explain why this
> 
> > (The test_helper::~test_helper() destructor calls ebpf_flush_epoch() that ends up blocking on a lock if called at DISPATCH.)
> 
> causes a crash with this stack? [#2973 (comment)](https://github.com/microsoft/ebpf-for-windows/issues/2973#issuecomment-1759363487)
> 
> (I thought we'd see a driver verifier bugcheck detecting the IRQL deadlock, rather than an access violation exception code.)

I was puzzled by the stack dump in [#2973 (comment)](https://github.com/microsoft/ebpf-for-windows/issues/2973#issuecomment-1759363487) as well, as I did not see any other instances in the subsequent dumps.   Also, by the time I got to investigating this issue those artifacts had expired so couldn't investigate further.  All the more puzzling is the fact that this is a _user_ _mode_ test (we can't do fault injection in kernel mode, and driver verifier won't kick in either) so it's odd/interesting that it can trigger a kernel mode crash.

I have focused instead on the far more consistent 'hang' as seen all subsequent crash dump artifacts.   Nonetheless, I'm keeping an eye open for the originally reported crash and dig in if it does repro again.

File: libs/runtime/unit/platform_unit_test.cpp (hunk 1 above)
Comment (dv-msft) [reply libs/runtime/unit/platform_unit_test.cpp:1214]: > Can we make this lower on scope exit?

Not sure I follow? There _is_ a `KeLowerIrql` call at the end of that test function, or did I misunderstand your comment?

**Review by rectified95 — APPROVED** [review]
LGTM
Don't forget to update the PR title and description to reflect the fix for the crash as well.

**Review by rectified95 — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**